.github/workflows/haskell-ci.yml

```diff
@@ -0,0 +1,208 @@
+name: Haskell CI
+
+defaults:
+  run:
+    shell: bash --noprofile --norc -euxo pipefail {0}
+
+on:
+  workflow_call:
+    inputs:
+      build-targets:
+        description: Cabal targets to build, e.g. `lib:my-lib`.
+        type: string
+        required: false
+        default: all
+      cabal:
+        description: Cabal version, e.g., "3.12.0.0"
+        type: string
+        required: false
+        default: latest
+      check:
+        description: Whether to run `cabal check`.
+        type: boolean
+        required: false
+        default: true
+      compat:
+        description: Whether to run the Cabal/GHC compatibility check.
+        type: boolean
+        required: false
+        default: true
+      compat-dirs:
+        description: Directories in which to run the Cabal/GHC compatibility check, relative to the repo root.
+        type: string
+        required: false
+        default: "."
+      configure-flags:
+        description: Flags to pass to `cabal configure`, e.g., `--ghc-options='-Werror=default -j'`
+        type: string
+        required: false
+        # For an explanation of `--ghc-options='-j'`, see Note [Parallelism].
+        default: --ghc-options='-j' --enable-tests
+      ghc:
+        description: GHC version, e.g., "9.10.1"
+        type: string
+        required: false
+        default: latest
+      haddock:
+        description: Whether to run `cabal haddock`.
+        type: boolean
+        required: false
+        default: true
+      os:
+        description: |
+          Runner OS, e.g., "ubuntu-latest".
+
+          Set this to the value of `runs-on`, *not* `runner.os`.
+        type: string
+        required: true
+      sdist:
+        description: |
+          Whether to run `cabal sdist`.
+
+          This will also cause later steps to be run in the unpacked sdist.
+        type: boolean
+        required: false
+        default: true
+      test:
+        description: Whether to run `cabal test`.
+        type: boolean
+        required: false
+        default: true
+      test-targets:
+        description: Cabal targets to test, e.g. `test:my-tests`.
+        type: string
+        required: false
+        default: all
+
+# General commentary
+# ------------------
+#
+# Steps that are less likely to fail should come later, because failing earlier
+# saves time and resources. For example, `cabal check` should come after `cabal
+# build`, because edits to the Cabal file are generally more rare than edits to
+# Haskell source files.
```

Comment:
This should probably advise that the `cabal configure` should be one of the first steps.  I know that seems obvious, but this CI might be edited by someone not familiar with intricacies of `cabal`.

Reply:
I don't know if this workflow *should* be modified by someone who isn't familiar with `cabal`, and if it is, it definitely should be *reviewed* by someone with knowledge of `cabal`. However, I have pushed a commit mentioning this to #18